### DISCUSSION TITLE
Writing scripts for CI/CD

### DISCUSSION THREAD

**Discussion by momotofu** [planetscale/discussion#124] [Q&A]
Hey 👋🏻, PlanetScale team and friends,

I've written some scripts using your CLI (great DevEx and design btw). So far the steps my scripts take are:
1. Create a unique test branch
2. Update an environment variable containing that branch name
3. Concurrently (using the npm package concurrently) establish a connection to the new branch
4. Concurrently pause for about 20 seconds (Is there a better way to do this?) to ensure there is a connection and then run a seed script (using Prisma)
5. Concurrently pause for about 20 seconds before starting my backend tests using jest
6. Once my tests are finished running, or I've stopped them (in case of watch) the scripts delete the unique test branch and all concurrent processes are stopped.

This works, but it's not the greatest since I need to pause scripts to make sure there is a db connection and pause to run migrations. All this before I can safely run tests. The time paused is lengthy, and variable since it would depend on the script runner's internet connection.

Does your team have any ideas on how I could improve this solution? Is there a "best practices" for integrating CI/CD using the PlanetScale CI?

Kind regards,
Christopher

**Comment by johnyerhot** [comment]
Hey @momotofu 👋 

> Concurrently pause for about 20 seconds (Is there a better way to do this?) to ensure there is a connection and then run a seed script (using Prisma)

I'm curious if trying to this would work better. Use the `-f json` flag with the CLI and some `jq`-fu to poll the status of the new branch to ensure it's ready before starting the seeding. Usually it's almost immediately ready upon being created, but that's not always the case as you've found.

You can use this to get the overall status of a given branch:

```
pscale branch show <database-nane> <branch-name> -f json

{
  "name": "name",
  "parent_branch": "parent",
  "region": {
    "slug": "us-east",
    "display_name": "US East",
    "location": "Northern Virginia",
    "enabled": true
  },
  "ready": true,
  "production": false,
  "created_at": "2021-11-02T21:14:34.114Z",
  "updated_at": "2021-11-02T21:14:49.128Z",
  "access_host_url": "<host>.us-east-4.psdb.cloud"
}
```

So something like this would return whether the branch is ready or not for connections:

```
pscale branch show <database> <branch> -f json | jq '.ready'
true
```

I would loop that until you get `true` back and then move to the next thing.

Is there a reason for using concurrently here? Most of the operations happening are fairly linear - create branch, once it's ready seed it, after that run tests. I'm not super familiar with the npm concurrently package but given my "read the npm page" understanding, but it seems like trying to do it all ... concurrently (lol, unintended pun 😆 ) may be adding more complexity than it's worth.

My thought is that a more simple and linear version without concurrently that moves to the next step or stops depending on the exit of the previous step is the way to go.

**Reply by momotofu** [reply]
Hey @johnyerhot 👋🏻,

> My thought is that a more simple and linear version without concurrently that moves to the next step or stops depending on the exit of the previous step is the way to go.

Thank you. Having the script be ran in a linear fashion is exactly what I'm wanting to achieve. I'm lacking some understanding though. Once my script opens a connection to the new branch, the scripts freeze. Or in other words, the rest of the steps stall after running `pscale connect my-database branch-name --port 3309`. I looked into running it in the background, but that still caused the rest of the file to stall.

That's why I crafted the concurrent + sleep approach. 

Can you explain why opening a connection pauses execution? What terms am I missing?

**Reply by ydennisy** [reply]
hey @momotofu I think you can do the following

```
pscale connect my-database branch-name --port 3309 &
```

_Added the_ `&`

Essentially you need to run that command in the background or a new process or it stops the execution.

**Comment by ydennisy** [comment]
Hi @momotofu if you could share more of your script / approach that would be great!

I am trying to figure out the best practices here too!

Are you using Prisma by any chance? Also how do you seed your DB?
